validator/validator.c

```diff
@@ -2449,10 +2451,51 @@ processFinished(struct module_qstate* qstate, struct val_qstate* vq,
 			return 0;
 		}
 
+		if(qstate->env->cfg->serve_expired &&
+			(e=msg_cache_lookup(qstate->env, qstate->qinfo.qname,
+			qstate->qinfo.qname_len, qstate->qinfo.qtype,
+			qstate->qinfo.qclass, qstate->query_flags,
+			0 /*now; allow expired*/,
+			1 /*wr; we may update the data*/))) {
+			struct reply_info* rep = (struct reply_info*)e->entry.data;
+			if(rep && rep->security > sec_status_bogus &&
+				(!qstate->env->cfg->serve_expired_ttl ||
+				 qstate->env->cfg->serve_expired_ttl_reset ||
+				*qstate->env->now <= rep->serve_expired_ttl)) {
+				verbose(VERB_ALGO, "validation failed but "
+					"previously cached valid response "
+					"exists; set serve-expired-norec-ttl "
+					"for response in cache");
+				rep->serve_expired_norec_ttl = NORR_TTL +
+					*qstate->env->now;
```

Comment:
Should this perhaps be the bogus ttl, `val-bogus-ttl`, since there is a validation failure perhaps it could wait for that before it retries validation.
```suggestion
				rep->serve_expired_norec_ttl = ve->bogus_ttl +
					*qstate->env->now;
```

Reply:
Maybe, and then clients will get the immediate cache response for the following "bogus_ttl" seconds, instead of waiting for resolution/validation first.
Validation will still happen in each such case since serve_expired_norec_ttl is essentially turning serve-expired-client-timeout configurations to just serve-expired (immediate cache answers and then resolution).

Reply:
I see, that resolved also the two other questions I had. This behaviour should maybe be noted at the header file where the serve_expired_norec_ttl variable is declared. In that case the bogus-ttl is not really that applicable, and the time change suggested does not make that much sense to apply.

Reply:
Just added more text on the header file

Reply:
That is nice to read up, thank you for the explanatory text.